academy/ibc-dev/ibcmw-develop.md

```diff
@@ -0,0 +1,456 @@
+
+The interface could be found [here](https://github.com/cosmos/ibc-go/blob/main/modules/core/05-port/types/module.go).
+
+Likewise when developing IBC middleware, the `Middleware` interface should be implemented. It can be found in the same file as the `IBCModule` and looks as follows.
+
+```go
+type Middleware interface {
+    IBCModule
+    ICS4Wrapper
+}
+```
+
+Note that middleware must implement `IBCModule` to wrap communication from core IBC to underlying application and `ICS4Wrapper` to wrap communication from underlying application to core IBC.
+
+The former we know from the IBC custom app section, the latter is an interface like so:
+
+```go
+// This is implemented by ICS4 and all middleware that are wrapping base application.
+// The base application will call `sendPacket` or `writeAcknowledgement` of the middleware directly above them
+// which will call the next middleware until it reaches the core IBC handler.
+type ICS4Wrapper interface {
+    SendPacket(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.Packet) error
+    WriteAcknowledgement(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.Packet, ack exported.Acknowledgement) error
+    GetAppVersion(ctx sdk.Context, portID, channelID string) (string, bool)
+}
+```
+
+When developing custom middleware, we can implement these interfaces for a new `IBCMiddleware` that has access to its keeper and an underlying application. In the `ibc_middleware.go` file:
+
+```go
+import (
+	...
+	porttypes "github.com/cosmos/ibc-go/v5/modules/core/05-port/types"
+	...
+)
+var _ porttypes.Middleware = &IBCMiddleware{}
+
+type IBCMiddleware struct {
+	app    porttypes.IBCModule
+	keeper keeper.Keeper
+}
+
+// IBCMiddleware creates a new IBCMiddleware given the associated keeper and underlying application
+func NewIBCMiddleware(app porttypes.IBCModule, k keeper.Keeper) IBCMiddleware {
+	return IBCMiddleware{
+		app:    app,
+		keeper: k,
+	}
+}
+```
+
+Below we'll take a closer look at how to implement the handshake callbacks and the packet callbacks, to satisfy the `IBCModule` interface. Furthermore also the `SendPacket`, `WriteAcknowledgement` and `GetAppVersion` methods satisfying the `ICS4Wrapper` interface.
+
+As a reminder, let's look once more at the diagram representing the information flow between core IBC and the base application with a middleware (stack) apllied.
+
+![middleware_stack](./images/placeholder_middleware_stack.svg)
+
+## Channel handshake callbacks
+
+The first type of callbacks from core IBC to the application are the channel handshake callbacks. When a middleware (stack) is applied, every piece of middleware will have access to the underlying application (base application or downstream middleware). For the handshake callbacks, function calls happen from core IBC down to base application via middleware. Each middleware will call the underlying application's callback but it can execute custom application logic before doing so.
```

Comment:
"from core IBC down to base application via middleware"

Does it make more sense to phrase this way?

> from core IBC through the middleware and then to base application.

This way the sentence structure also follows the execution path? Bit of a nit. I don't know if you want to also explain that the calls go from the top-level middleware to lowest middleware then to baseapp in here or if you want to explain this elsewhere

Reply:
> This way the sentence structure also follows the execution path? 
That makes sense to me and is more intuitive indeed.

The direction of the calls would be explained in the intro already (the readme files here do not show up in order) and just reminded on the diagram